### PR TITLE
CASSANDRA-17402: Fix ObjectSizes implementation and usages

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

File: src/java/org/apache/cassandra/utils/ObjectSizes.java [hunk 1]
```diff
@@ -96,58 +112,97 @@ private static long sizeOfArray(int length, long elementSize)
...
-        if (buffer.capacity() > buffer.remaining())
-            return buffer.remaining();
-        return BUFFER_EMPTY_SIZE + sizeOfArray(buffer.capacity(), 1);
-    }
+            return DIRECT_BUFFER_HEAP_SIZE;
 
-    public static long sizeOfEmptyHeapByteBuffer()
-    {
-        return BUFFER_EMPTY_SIZE;
+        int arrayLen = buffer.array().length;
```
Comment (blambov) [inline src/java/org/apache/cassandra/utils/ObjectSizes.java:158]: Great idea!

**Review by blambov — APPROVED** [review]

File: src/java/org/apache/cassandra/utils/ObjectSizes.java (hunk 1 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/utils/ObjectSizes.java:158]: thanks

File: src/java/org/apache/cassandra/utils/ObjectSizes.java [hunk 2]
```diff
@@ -96,58 +112,97 @@ private static long sizeOfArray(int length, long elementSize)
...
     }
 
     /**
      * Memory a String consumes
+     *
      * @param str String to calculate memory size of
      * @return Total in-memory size of the String
      */
-    //@TODO hard coding this to 2 isn't necessarily correct in Java 11
+    // TODO hard coding this to 2 isn't necessarily correct in Java 11
```
Comment (adelapena) [inline src/java/org/apache/cassandra/utils/ObjectSizes.java:199]: Is this TODO comment still valid?

Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/utils/ObjectSizes.java:199]: yes, I didn't change that logic at all

Comment (adelapena) [reply src/java/org/apache/cassandra/utils/ObjectSizes.java:199]: Well, the comment is mentioning the hardcoded "2" value, I understand that meaning the 2 bytes that were passed to `MemoryLayoutSpecification.sizeOfArray`. However, this has changed to send `Character.SIZE`, which has a value of 16 bits. So we are calling `sizeOfArray` with the size of a char in bits, when it expects bytes. I guess we should change it to `Character.BYTES`.

File: test/unit/org/apache/cassandra/db/CellSpecTest.java [hunk 3]
```diff
@@ -39,6 +39,7 @@
 import org.apache.cassandra.db.rows.NativeCell;
 import org.apache.cassandra.schema.ColumnMetadata;
 import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
```
Comment (adelapena) [inline test/unit/org/apache/cassandra/db/CellSpecTest.java:42]: Nit: unused, can be fixed on commit

**Pull request closed without merging**